File: frontend/src/api/csr.ts
import { get, post, del } from './request';
import type { CSRRecord, PageResponse } from '@/types';

export interface GenerateCSRRequest {
  commonName: string;
  sans: string[];
  keyAlgorithm: 'RSA' | 'ECC';
  keySize: number;
  // 阿里云 CreateCsr 参数
  countryCode: string;     // 国家代码，如 CN
  province: string;        // 省份
  locality: string;        // 城市
  corpName?: string;       // 单位名称（可选）
  department?: string;     // 部门（可选）
}

export interface CSRInfo {
  commonName: string;
  sans: string[];
  keyAlgorithm: string;
  keySize: number;
}

export interface CSRListParams {
  page: number;
  pageSize: number;
  search?: string;
}

// 生成 CSR
export function generateCSR(data: GenerateCSRRequest): Promise<CSRRecord> {
  return post<CSRRecord>('/csrs/generate', data);
}

// 获取 CSR 列表
export function getCSRList(params: CSRListParams): Promise<PageResponse<CSRRecord>> {
  return get<PageResponse<CSRRecord>>('/csrs', { params });
}

// 获取单个 CSR
export function getCSR(id: number): Promise<CSRRecord> {
  return get<CSRRecord>(`/csrs/${id}`);
}

// 删除 CSR
export function deleteCSR(id: number): Promise<void> {
  return del<void>(`/csrs/${id}`);
}

// 解析 CSR
export function parseCSR(csrPEM: string): Promise<CSRInfo> {
  return post<CSRInfo>('/csrs/parse', { csrPEM });
}

// 下载 CSR
export function downloadCSR(id: number): Promise<Blob> {
  return get<Blob>(`/csrs/${id}/download-csr`, {
    responseType: 'blob',
  });
}

// 下载私钥
export function downloadPrivateKey(id: number): Promise<Blob> {
  return get<Blob>(`/csrs/${id}/download-key`, {
    responseType: 'blob',
  });
}

File: backend/internal/model/csr.go
package model

// CSRRecord CSR 记录模型
type CSRRecord struct {
	BaseModel
	CommonName          string `gorm:"size:255;not null" json:"common_name"`
	SAN                 string `gorm:"size:1000" json:"san"` // 逗号分隔的域名列表
	KeyAlgorithm        string `gorm:"size:50;not null" json:"key_algorithm"`
	KeySize             int    `json:"key_size"`
	CSRPEM              string `gorm:"type:text" json:"-"`
	PrivateKeyEncrypted string `gorm:"type:text" json:"-"`
	Status              string `gorm:"size:50;not null" json:"status"`
	// 阿里云 CreateCsr 参数
	CountryCode string `gorm:"size:10" json:"country_code"`
	Province    string `gorm:"size:100" json:"province"`
	Locality    string `gorm:"size:100" json:"locality"`
	CorpName    string `gorm:"size:255" json:"corp_name"`
	Department  string `gorm:"size:100" json:"department"`
}

// TableName 指定表名
func (CSRRecord) TableName() string {
	return "csr_records"
}
